Create diagnosticsource-getting-started.md

mikelle-rogers opened the pull request:
Fixes [#49082](https://github.com/dotnet/runtime/issues/49082) and fixes [#3029](https://github.com/dotnet/diagnostics/issues/3029)

mikelle-rogers committed:
Create diagnosticsource-getting-started.md

This is a rough draft, all suggestions appreciated.

noahfalk commented on docs/core/diagnostics/diagnosticsource-getting-started.md:
```diff
@@ -0,0 +1,27 @@
+
+**This article applies to: ✔️** .NET Core 3.1 and later versions **✔️** .NET Framework 4.5 and later versions
+
+This walkthrough shows <xref:System.Diagnostics.DiagnosticSource?displayProperty=nameWithType> 
```
Do we expect readers to understand what the purpose of DiagnosticSource is at this point? In the EventSource docs there was an 'overview' page before the 'getting started' page that had a paragraph describing what EventSource was and what it could be useful for. We don't have to rigidly follow the same structure, but somewhere very early in the documentation devs will probably want to understand this.

noahfalk commented on docs/core/diagnostics/diagnosticsource-getting-started.md:
```diff
@@ -0,0 +1,27 @@
+the `DiagnosticListener` implements. `DiagnosticListeners` are given names when they are created
+and this name should be the name of logical grouping of related events (typically the component).
+Later this name is used to find the Listener and subscribe to any of its events. `DiagnosticListeners` have a name, which is used to represent the component associated with the event.
+Thus the event names only need to be unique within a component.
```
I suggest any getting started material should cover both emitting the event and listening to the event. This should make it more concrete how the feature works end-to-end and why it could be useful.

mikelle-rogers replied:
@noahfalk may we please meet up and discuss this?

noahfalk replied:
Certainly! Sorry I've had meetings throughout the afternoon and now about to go eat.

mikelle-rogers committed:
DiagnosticSource Overview page

mikelle-rogers committed:
Alter formatting and clarify contents of the file

mikelle-rogers committed:
Update diagnosticsource.md

mikelle-rogers committed:
Fixing lint failures

gfoidl commented on docs/core/diagnostics/diagnosticsource-instrumentation-consumption.md:
```diff
@@ -0,0 +1,263 @@
+
+## Instrumenting with DiagnosticSource/DiagnosticListener
+
+We show how to instrument DiagnosticListener using the DiagnosticListener we made in the [Getting Started Guide](/docs/core/diagnostics/diagnosticsource-getting-started).
```
```suggestion
We show how to instrument DiagnosticListener using the DiagnosticListener we made in the [Getting Started Guide](/docs/core/diagnostics/diagnosticsource-getting-started.md).
```

gfoidl commented on docs/core/diagnostics/diagnosticsource-instrumentation-consumption.md:
```diff
@@ -0,0 +1,263 @@
+Architectural elements shown in the above code are as follows.
+Every event has a `string` name (e.g. `RequestStart`), and exactly one `object` as a payload.
+If you need to send more than one item, you can do so by creating an `object` with all information
+in it as properties. C#'s [anonymous type](https://msdn.microsoft.com/library/bb397696.aspx)
```
```suggestion
in it as properties. C#'s [anonymous type](https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/types/anonymous-types)
```
Or is there a linkage in-between the docs that suits better?
The MSDN link is outdated.

gfoidl commented on docs/core/diagnostics/diagnosticsource-instrumentation-consumption.md:
```diff
@@ -0,0 +1,263 @@
+`IObservable<DiagnosticListener>`.
+
+The `IObservable` interface is the 'callback' version of the `IEnumerable` interface. You can learn
+more about it at the [Reactive Extensions](https://msdn.microsoft.com/data/gg577609.aspx) site.
```
```suggestion
more about it at the [Reactive Extensions](https://docs.microsoft.com/en-us/previous-versions/dotnet/reactive-extensions/hh242985) site.
```
Same here.

gfoidl commented on docs/core/diagnostics/diagnosticsource-instrumentation-consumption.md:
```diff
@@ -0,0 +1,263 @@
+option if the payloads were generated using anonymous types. This overhead can be reduced by
+making fast, specialized property fetchers either using `PropertyInfo.CreateDelegate` or
+`ReflectEmit`, but that is beyond the scope of this document.
+(See the [PropertySpec](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L784)
```
This links points anywhere in that file: 
![grafik](https://user-images.githubusercontent.com/5755834/166887943-bcd59810-55cb-45c7-a5f6-a2d7d5dfdca3.png)

Instead of _main_-branch this should refer to a commit-hash, so the link doesn't get outdated with each update on that file.

mikelle-rogers replied:
@noahfalk where should this link be linked to?

noahfalk replied:
This is probably the reference that was intended: https://github.com/dotnet/runtime/blob/6de7147b9266d7730b0d73ba67632b0c198cb11e/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L1235

mikelle-rogers committed:
Fix up the code and finish up the documentation

mikelle-rogers committed:
Not needed

mikelle-rogers committed:
Update the date

mikelle-rogers committed:
Combine into one file

mikelle-rogers committed:
Not needed

mikelle-rogers committed:
Remove whitespace

noahfalk commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,326 @@
+of the C# language require all the work of creating the payload `object` and calling `Write()` to be
+done, even though nothing is actually listening for the data. By guarding the `Write()` call, we
+make it efficient when the source is not enabled.
+
```
You might to introduce the code you are working with at the top of the section and then explain it as you go. To make this example a little more realistic we should probably change the naming. In practice a developer would be starting with some code that does some useful work and then they modify that code to add logging. The names of the class and method would still be based on the original work the class was doing.

```C#
// before adding logging
class HttpClient
{
    byte[] SendWebRequest(string url)
    {
        // pretend this sends an HTTP request to the url and gets back a reply
        byte[] reply = new byte[] {};
        return reply;
}
    
// after adding logging
class HttpClient
{
    private static DiagnosticSource logger= new DiagnosticListener("System.Net.Http");
    public byte[] SendWebRequest(string url)
    {
        if (logger.IsEnabled("RequestStart"))
        {
            httpLogger.Write("RequestStart", new { Url = url });
        }
        // pretend this sends an HTTP request to the url and gets back a reply
        byte[] reply = new byte[] {};
        return reply;
    }
}
```

noahfalk commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,326 @@
+            // Here is where we put code to subscribe to the Listener.
+        }
+    };
+    IObserver<DiagnosticListener> TheIObserver = new Observer<DiagnosticListener>(onNewListener, null);
```
```suggestion
    IObserver<DiagnosticListener> observer = new Observer<DiagnosticListener>(onNewListener, null);
```

Nit: .NET C# coding conventions recommend using camelCase for variables.

noahfalk commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,326 @@
+        Console.WriteLine("From Listener {0} Received Event {1} with payload {2}", networkListener.Name, evnt.Key, evnt.Value.ToString());
+
+    // Turn it into an observer (using System.Reactive.Core's AnonymousObserver)
+    Observer<KeyValuePair<string, object>> observer = new AnonymousObserver<KeyValuePair<string, object>>(callback);
```
We already defined our own observer so maybe keep using it rather than switching to the one in System.Reactive?

mikelle-rogers committed:
Fixed suggestions from code review

mikelle-rogers committed:
Two additional edits

tommcdon reviewed (approved):
LGTM thanks!

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+
+Like all calls to `Subscribe()`, this one returns an `IDisposable` that represents the subscription itself.
+Callbacks will continue to happen as long as nothing calls `Dispose()` on this subscription object.
+The above code never calls `Dispose()`, so it will receive callbacks forever.
```
```suggestion
The code example never calls `Dispose()`, so it will receive callbacks forever.
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+
+When you subscribe to `AllListeners`, you get a callback for ALL ACTIVE `DiagnosticListeners`.
+Thus, upon subscribing, you get a flurry of callbacks for all existing `DiagnosticListeners`, and as new ones
+are created, you receive a callback for those as well. You receive a complete list of everything it is possible
```
```suggestion
are created, you receive a callback for those as well. You receive a complete list of everything it's possible
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+are created, you receive a callback for those as well. You receive a complete list of everything it is possible
+to subscribe to.
+
+#### Subscribing to DiagnosticListeners
```
```suggestion
#### Subscribe to DiagnosticListeners
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+#### Subscribing to DiagnosticListeners
+
+A `DiagnosticListener` implements the `IObservable<KeyValuePair<string, object>>` interface, so you can
+call `Subscribe()` on it as well. We show how to fill out the previous example:
```
```suggestion
call `Subscribe()` on it as well. The following code shows how to fill out the previous example:
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
````diff
@@ -0,0 +1,343 @@
+    // At some point you may wish to dispose the networkSubscription.
+```
+
+In the above example, after finding the 'System.Net.Http' `DiagnosticListener`, an action is created that
````
```suggestion
In this example, after finding the 'System.Net.Http' `DiagnosticListener`, an action is created that
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
````diff
@@ -0,0 +1,343 @@
+```
+
+In the above example, after finding the 'System.Net.Http' `DiagnosticListener`, an action is created that
+prints out the name of the listener, event, and `payload.ToString()`. Please note the following.
````
```suggestion
prints out the name of the listener, event, and `payload.ToString()`.
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+In the above example, after finding the 'System.Net.Http' `DiagnosticListener`, an action is created that
+prints out the name of the listener, event, and `payload.ToString()`. Please note the following.
+
+`DiagnosticListener` implements `IObservable<KeyValuePair<string, object>>`. This means
```
```suggestion
> [!NOTE]
> `DiagnosticListener` implements `IObservable<KeyValuePair<string, object>>`. This means
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+
+`DiagnosticListener` implements `IObservable<KeyValuePair<string, object>>`. This means
+ on each callback we get a `KeyValuePair`. The key of this pair is the name of the event
+ and the value is the payload `object`. In the code above we simply log this information
```
```suggestion
 and the value is the payload `object`. The code example simply logs this information
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+ and the value is the payload `object`. In the code above we simply log this information
+ to the console.
+
+It is important to keep track of subscriptions to the `DiagnosticListener`. In the above code the
```
```suggestion
It's important to keep track of subscriptions to the `DiagnosticListener`. In the previous code, the
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+ to the console.
+
+It is important to keep track of subscriptions to the `DiagnosticListener`. In the above code the
+networkSubscription variable remembers this. If another `creation` is formed, one must
```
```suggestion
`networkSubscription` variable remembers this. If you form another `creation`, you must
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+
+The `DiagnosticSource`/`DiagnosticListener` code is thread safe, but the
+callback code also needs to be thread safe. To ensure the callback code is thread safe, locks are used. It is possible to create two `DiagnosticListeners`
+with the same name at the same time. To avoid races updates of shared variables are performed under the protection of a lock.
```
```suggestion
with the same name at the same time. To avoid race conditions, updates of shared variables are performed under the protection of a lock.
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+callback code also needs to be thread safe. To ensure the callback code is thread safe, locks are used. It is possible to create two `DiagnosticListeners`
+with the same name at the same time. To avoid races updates of shared variables are performed under the protection of a lock.
+
+Once the above code is run, the next time a `Write()` is done on 'System.Net.Http' `DiagnosticListener`
```
```suggestion
Once the previous code is run, the next time a `Write()` is done on 'System.Net.Http' `DiagnosticListener`
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+the information will be logged to the console.
+
+Subscriptions are independent of one another. As a result, other code
+can do exactly the same thing as the code above, and generate two 'pipes' of the logging
```
```suggestion
can do exactly the same thing as the code example and generate two 'pipes' of the logging
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+can do exactly the same thing as the code above, and generate two 'pipes' of the logging
+information.
+
+#### Decoding Payloads
```
```suggestion
#### Decode payloads
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+The `KeyvaluePair` that is passed to the callback has the event name and payload, but the payload is typed simply as
+an `object`. There are two ways of getting more specific data:
+
+1. If the payload is a well known type (e.g. a `string`, or an `HttpMessageRequest`) then you can simply
```
```suggestion
- If the payload is a well known type (for example, a `string` or an `HttpMessageRequest`), then you can simply
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+   cast the `object` to the expected type (using the `as` operator so as not to cause an exception if
+   you are wrong) and then access the fields. This is very efficient.
+
+2. Use reflection API. For example, assume the method below is present.
```
```suggestion
- Use reflection API. For example, assume the following method is present.
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
````diff
@@ -0,0 +1,343 @@
+    }
+```
+
+The `listener.Subscribe()` call above could be replaced with the following code, to decode the payload more fully.
````
```suggestion
To decode the payload more fully, you could replace the `listener.Subscribe()` call with the following code.
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+option if the payloads were generated using anonymous types. This overhead can be reduced by
+making fast, specialized property fetchers either using `PropertyInfo.CreateDelegate` or
+`ReflectEmit`, but that is beyond the scope of this document.
+(See the [PropertySpec](https://github.com/dotnet/runtime/blob/6de7147b9266d7730b0d73ba67632b0c198cb11e/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L1235)
```
```suggestion
(For an example of a fast, delegate-based property fetcher, see the [PropertySpec](https://github.com/dotnet/runtime/blob/6de7147b9266d7730b0d73ba67632b0c198cb11e/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L1235)
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+making fast, specialized property fetchers either using `PropertyInfo.CreateDelegate` or
+`ReflectEmit`, but that is beyond the scope of this document.
+(See the [PropertySpec](https://github.com/dotnet/runtime/blob/6de7147b9266d7730b0d73ba67632b0c198cb11e/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L1235)
+class used in the `DiagnosticSourceEventSource` for an example of a fast, delegate-based property fetcher.)
```
```suggestion
class used in the `DiagnosticSourceEventSource`.)
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+
+Note that using reflection is relatively expensive. However, using reflection is the only
+option if the payloads were generated using anonymous types. This overhead can be reduced by
+making fast, specialized property fetchers either using `PropertyInfo.CreateDelegate` or
```
```suggestion
making fast, specialized property fetchers using either `PropertyInfo.CreateDelegate` or
```

gewarren replied:
I'm not seeing either of these two methods in the .NET API browser?

josalem replied:
Based on the context, I'm guessing it means doing something like
```csharp
PropertyInfo pi = ...;
pi.GetMethod.CreateDelegate(...)
```

Where `CreateDelegate` is referring to `MethodInfo.CreateDelegate`.

e.g., [its usage in `DiagnosticSourceEventSoruce`.](https://source.dot.net/#System.Diagnostics.DiagnosticSource/System/Diagnostics/DiagnosticSourceEventSource.cs,1392)

mikelle-rogers replied:
@noahfalk are you aware of why CreateDelegate and ReflectEmit are not listed in the PropertyInfo [page](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.propertyinfo?view=net-6.0)?

noahfalk replied:
As far as I know those methods never existed. Probably what Vance (the original author of this doc) meant was propInfo.GetMethod.CreateDelegate() as @josalem was suggesting.

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+Note that using reflection is relatively expensive. However, using reflection is the only
+option if the payloads were generated using anonymous types. This overhead can be reduced by
+making fast, specialized property fetchers either using `PropertyInfo.CreateDelegate` or
+`ReflectEmit`, but that is beyond the scope of this document.
```
```suggestion
`ReflectEmit`, but that's beyond the scope of this article.
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+#### Filtering
+
+In the example above, the code uses the `IObservable.Subscribe()` method to hook up the callback. This
+causes all events to be given to the callback. However `DiagnosticListener` has overloads of
```
```suggestion
causes all events to be given to the callback. However, `DiagnosticListener` has overloads of
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
````diff
@@ -0,0 +1,343 @@
+    // subscription.Dispose() to stop the callbacks.
+```
+
+This very efficiently subscribes to only the 'RequestStart' events. All other events will cause the `DiagnosticSource.IsEnabled()`
````
```suggestion
This efficiently subscribes to only the 'RequestStart' events. All other events will cause the `DiagnosticSource.IsEnabled()`
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
````diff
@@ -0,0 +1,343 @@
+```
+
+This very efficiently subscribes to only the 'RequestStart' events. All other events will cause the `DiagnosticSource.IsEnabled()`
+method to return `false`, and thus be efficiently filtered out.
````
```suggestion
method to return `false` and thus be efficiently filtered out.
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+This very efficiently subscribes to only the 'RequestStart' events. All other events will cause the `DiagnosticSource.IsEnabled()`
+method to return `false`, and thus be efficiently filtered out.
+
+##### Context-based Filtering
```
```suggestion
##### Context-based filtering
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+##### Context-based Filtering
+
+Some scenarios require advanced filtering based on extended context.
+Producers may call `DiagnosticSource.IsEnabled()` overloads and supply additional event properties as shown in the code below.
```
```suggestion
Producers can call <xref:System.Diagnostics.DiagnosticSource.IsEnabled%2A?displayProperty=nameWithType> overloads and supply additional event properties as shown in the following code.
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
````diff
@@ -0,0 +1,343 @@
+        httpLogger.Write("RequestStart", new { Url="http://clr", Request=aRequest });
+```
+
+The next code demonstrates that consumers may use such properties to filter events more precisely:
````
```suggestion
The next code example demonstrates that consumers can use such properties to filter events more precisely.
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
````diff
@@ -0,0 +1,343 @@
+    IDisposable subscription = listener.Subscribe(observer, predicate);
+```
+
+Note that producers are not aware of the filter a consumer has provided. `DiagnosticListener`
````
```suggestion
Producers are not aware of the filter a consumer has provided. `DiagnosticListener`
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+Note that producers are not aware of the filter a consumer has provided. `DiagnosticListener`
+will invoke the provided filter, omitting additional arguments if necessary, thus the filter
+should expect to receive a `null` context.
+Producers should enclose `IsEnabled()` calls with event name and context with pure `IsEnabled()`
```
```suggestion
If a producer calls `IsEnabled()` with event name and context, those calls are enclosed in an overload that takes
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+will invoke the provided filter, omitting additional arguments if necessary, thus the filter
+should expect to receive a `null` context.
+Producers should enclose `IsEnabled()` calls with event name and context with pure `IsEnabled()`
+calls for event name, so consumers must ensure that their filter allows events without context
```
```suggestion
only the event name. Consumers must ensure that their filter allows events without context
```

gewarren replied:
```suggestion
only the event name, so consumers must ensure that their filter allows events without context
```

gewarren reviewed (commented):
Some general guidance that explains a lot of my comments:

- Okay to use "you" but not "we" or "our".
- Shorten sentences to help with localization.
- Use active voice over passive.

josalem commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+    {
+        if (listener.Name == "System.Net.Http")
+        {
+          lock(allListeners)
```
Any locking in prescribed code snippets should be explained.

mikelle-rogers replied:
@josalem There is something in the text that describes the locking. Do you think something more than that is needed?

josalem replied:
I would refactor this snippet to include the _entire_ class definition, so it is obvious where `allListeners` comes from. Right now, it is a variable that isn't defined in the scope of the snippet.

josalem commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
````diff
@@ -0,0 +1,343 @@
+    // Typically you leave the listenerSubscription subscription active forever.
+    // However when you no longer want your callback to be called, you can
+    // call listenerSubscription.Dispose() to cancel your subscription to the IObservable.
+```
````
I think it would be better for teaching purposes to have a single code snippet with all the features implemented and commented. You can call out specific parts as you go, but having the implementation spread out as it is makes it difficult to follow and, more importantly, makes it difficult for users to copy/paste as a starting template for their own implementation.

mikelle-rogers committed:
Response to suggestions

Use you instead of we, shorten sentences to help with localization, use active voice, comments should be capitalized and have punctuation.

josalem commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
````diff
@@ -0,0 +1,343 @@
+
+Some scenarios require advanced filtering based on extended context.
+Producers may call `DiagnosticSource.IsEnabled()` overloads and supply additional event properties as shown in the code below.
+
+```C#
+    if (httpLogger.IsEnabled("RequestStart", aRequest, anActivity))
+        httpLogger.Write("RequestStart", new { Url="http://clr", Request=aRequest });
+```
````
A comment saying that `aRequest` and `anActivity` are the _current_ request and activity about to logged might make this clearer.

It's also worth adding some comments about how this is instrument site specific, i.e., users need to check the instrumentation site to see what types are being passed in to `IsEnabled3Args` so they know what to cast to.

josalem commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+    {
+        if (eventName == "RequestStart")
+        {
+            HttpRequestMessage request = context as HttpRequestMessage;
+            if (request != null)
+            {
+                return IsUriEnabled(request.RequestUri);
+            }
```
```suggestion
            if (context is HttpRequestMessage request)
            {
                return IsUriEnabled(request.RequestUri);
            }
```

josalem reviewed (commented):
In general, I think we should consolidate the code samples into all-inclusive samples and then break down their contents. Users will want to be able to visit this guide and leave with a snippet of code they can build on. Having everything broken up makes that a complex task of grokking the whole page.

noahfalk commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,343 @@
+active in the system at runtime. The API to accomplish this is the `AllListeners`
+`IObservable<DiagnosticListener>`.
+
+Implement an `Observer<T>` class that inherits from the `IObservable` interface, which is the 'callback' version of the `IEnumerable` interface. You can learn more about it at the [Reactive Extensions](https://docs.microsoft.com/dotnet/csharp/fundamentals/types/anonymous-types) site.
```
https://docs.microsoft.com/en-us/previous-versions/dotnet/reactive-extensions/hh242985(v=vs.103) this is the best text document on msdn I could find. I think any of these links would be fine.

mikelle-rogers committed:
Update to working example

Add clarifying comments, URLs for the APIs and responded to suggestions.

mikelle-rogers committed:
Add a line, remove a space

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
````diff
@@ -0,0 +1,411 @@
+
+Running the provided implementation prints to the console.
+
+```
````
```suggestion
```console
```

gewarren reviewed (commented):
Can you change all the \`\`\`C# lines to \`\`\`csharp?
And for all the output code, they should be labeled as `console`.

mikelle-rogers committed:
Updating from feedback on code review

gewarren reviewed (approved):
Please also add the file to the table of contents (toc.yml). Sorry I didn't catch that earlier.

mikelle-rogers commented:
> Please also add the file to the table of contents (toc.yml). Sorry I didn't catch that earlier.

Only this file, or the entire diagnostics repo? Is this the toc.yml file you mentioned? https://github.com/dotnet/docs/blob/dev/mirogers/diagnosticsource-getting-started/docs/toc.yml

mikelle-rogers commented:
> Please also add the file to the table of contents (toc.yml). Sorry I didn't catch that earlier.

Only this file, or the entire diagnostics repo? Is this the toc.yml file you mentioned? https://github.com/dotnet/docs/blob/dev/mirogers/diagnosticsource-getting-started/docs/toc.yml

gewarren commented:
> @gewarren
> 
> > Please also add the file to the table of contents (toc.yml). Sorry I didn't catch that earlier.
> 
> Only this file, or the entire diagnostics repo? Is this the toc.yml file you mentioned? https://github.com/dotnet/docs/blob/dev/mirogers/diagnosticsource-getting-started/docs/toc.yml

Just this file, since it's new. The TOC file is at https://github.com/dotnet/docs/blob/main/docs/fundamentals/toc.yml.

mikelle-rogers committed:
Add DiagnosticSource and DiagnosticListener to toc

mikelle-rogers committed:
Add .csproj and .cs files

mikelle-rogers committed:
Add links to the .cs and .csproj folders

mikelle-rogers committed:
Merge branch 'main' into dev/mirogers/diagnosticsource-getting-started

josalem commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
````diff
@@ -0,0 +1,413 @@
+{
+    IDisposable networkSubscription;
+    IDisposable listenerSubscription;
+    private readonly object allListeners = new();
+    public void Listening()
+    {
+        Action<KeyValuePair<string, object>> whenHeard = delegate (KeyValuePair<string, object> data)
+        {
+            Console.WriteLine($"Data received: {data.Key}: {data.Value}");
+        };
+        Action<DiagnosticListener> onNewListener = delegate (DiagnosticListener listener)
+        {
+            Console.WriteLine($"New Listener discovered: {listener.Name}");
+            //Suscribe to the specific DiagnosticListener of interest.
+            if (listener.Name == "System.Net.Http")
+            {
+                //Use lock to ensure the callback code is thread safe.
+                lock(allListeners)
+                {
+                    if (networkSubscription != null)
+                    {
+                        networkSubscription.Dispose();
+                    }
+                    IObserver<KeyValuePair<string, object>> iobserver = new Observer<KeyValuePair<string, object>>(whenHeard, null);
+                    networkSubscription = listener.Subscribe(iobserver);
+                }
+                
+            }
+        };
+        //Subscribe to discover all DiagnosticListeners
+        IObserver<DiagnosticListener> observer = new Observer<DiagnosticListener>(onNewListener, null);
+        //When a listener is created, invoke the onNext function which calls the delegate.
+        listenerSubscription = DiagnosticListener.AllListeners.Subscribe(observer);
+
+    }
+    // Typically you leave the listenerSubscription subscription active forever.
+    // However when you no longer want your callback to be called, you can
+    // call listenerSubscription.Dispose() to cancel your subscription to the IObservable.
+}
+```
````
I don't think it's necessary, but you could simplify this now that you've checked in the snippet as a separate program.

It would require two things:

1. Move the checked in .cs/.csproj pair to `docs/core/diagnostics/snippets/diagnosticsource/csharp`.
2. Add `// <WholeProgram>` to the beginning of the program and `// </WholeProgram>` at the end of the program.


Then you could replace this whole code snippet with:
```md
:::code language="csharp" source="snippets/diagnosticsource/csharp/Program.cs" id="WholeProgram":::
```

For any other snippets out of this code that you want to display, you just need to wrap the lines with
```csharp
// <Snippet1>
    var thisCodeIsInSnippet1 = 0;
// <Snippet 2>
    var thisCodeIsInSnippet1AndSnippet2 = 1;
// </Snippet2>
// </Snippet1>
```

Then, you could replace those code blocks with
```md
:::code language="csharp" source="snippets/diagnosticsource/csharp/Program.cs" id="Snippet1":::
:::code language="csharp" source="snippets/diagnosticsource/csharp/Program.cs" id="Snippet2":::
```

Again, I don't think it is absolutely necessary, but I thought I'd lay out how to do it.

gewarren replied:
Agreed - we shouldn't duplicate the code inline and in a separate file.

gewarren commented on docs/core/diagnostics/DiagnosticSource.csproj:
```diff
@@ -0,0 +1,10 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Can you put this and the .cs file into a subdirectory named "snippets"? That's the convention we use throughout this repo.

gewarren commented on docs/fundamentals/toc.yml:
```diff
@@ -654,6 +654,10 @@ items:
+      - name: DiagnosticSource and DiagnosticListener
+        items:
+        - name: Getting started
+          href: ../core/diagnistics/diagnosticsource-diagnosticlistener.md
```
```suggestion
          href: ../core/diagnostics/diagnosticsource-diagnosticlistener.md
          displayName: diagnosticsource, diagnosticlistener
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,413 @@
+ to the console.
+
+It's important to keep track of subscriptions to the `DiagnosticListener`. In the previous code, the
+'networkSubscription' variable remembers this. If you form another `creation`, you must
```
```suggestion
`networkSubscription` variable remembers this. If you form another `creation`, you must
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,413 @@
+the information will be logged to the console.
+
+Subscriptions are independent of one another. As a result, other code
+can do exactly the same thing as the code example, and generate two 'pipes' of the logging
```
```suggestion
can do exactly the same thing as the code example and generate two 'pipes' of the logging
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,413 @@
+option if the payloads were generated using anonymous types. This overhead can be reduced by
+making fast, specialized property fetchers using either [`PropertyInfo.GetMethod.CreateDelegate`](/api/system.reflection.methodinfo?view=net-6.0#:~:text=CreateDelegate(Type)) or
+[`System.Reflect.Emit`Namespace](/api/system.reflection.emit?view=net-6.0), but that's beyond the scope of this article.
+(For an example of a fast, delegate-based property fetcher, see [PropertySpec](https://github.com/dotnet/runtime/blob/6de7147b9266d7730b0d73ba67632b0c198cb11e/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L1235)
```
```suggestion
(For an example of a fast, delegate-based property fetcher, see the [PropertySpec](https://github.com/dotnet/runtime/blob/6de7147b9266d7730b0d73ba67632b0c198cb11e/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L1235)
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,413 @@
+Note that using reflection is relatively expensive. However, using reflection is the only
+option if the payloads were generated using anonymous types. This overhead can be reduced by
+making fast, specialized property fetchers using either [`PropertyInfo.GetMethod.CreateDelegate`](/api/system.reflection.methodinfo?view=net-6.0#:~:text=CreateDelegate(Type)) or
+[`System.Reflect.Emit`Namespace](/api/system.reflection.emit?view=net-6.0), but that's beyond the scope of this article.
```
```suggestion
the <xref:System.Reflection.Emit> namespace, but that's beyond the scope of this article.
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,413 @@
+
+Note that using reflection is relatively expensive. However, using reflection is the only
+option if the payloads were generated using anonymous types. This overhead can be reduced by
+making fast, specialized property fetchers using either [`PropertyInfo.GetMethod.CreateDelegate`](/api/system.reflection.methodinfo?view=net-6.0#:~:text=CreateDelegate(Type)) or
```
```suggestion
making fast, specialized property fetchers using either [PropertyInfo.GetMethod.CreateDelegate()](xref:System.Reflection.MethodInfo.CreateDelegate%2A) or
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,413 @@
+
+#### Filtering
+
+In the example above, the code uses the `IObservable.Subscribe()` method to hook up the callback. This
```
```suggestion
In the previous example, the code uses the `IObservable.Subscribe()` method to hook up the callback. This
```

gewarren commented on docs/core/diagnostics/diagnosticsource-diagnosticlistener.md:
```diff
@@ -0,0 +1,413 @@
+##### Context-based filtering
+
+Some scenarios require advanced filtering based on extended context.
+Producers can call xref:System.Diagnostics.DiagnosticSource.IsEnabled()%2A?displayProperty=nameWithType> overloads and supply additional event properties as shown in the following code.
```
```suggestion
Producers can call <xref:System.Diagnostics.DiagnosticSource.IsEnabled%2A?displayProperty=nameWithType> overloads and supply additional event properties as shown in the following code.
```

mikelle-rogers committed:
Move the file to the correct place

mikelle-rogers committed:
move location, add blocking

mikelle-rogers committed:
Moving to correct address for real

mikelle-rogers committed:
Linking the code to the snippits

mikelle-rogers committed:
Correct filepath

mikelle-rogers committed:
Responding to feedback

mikelle-rogers committed:
Added missing character

mikelle-rogers committed:
removing links
